File: subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/enumtype/EnumLitExpr.java
package hu.bme.mit.theta.core.type.enumtype;

import com.google.common.base.Objects;
import hu.bme.mit.theta.core.model.Valuation;
import hu.bme.mit.theta.core.type.LitExpr;
import hu.bme.mit.theta.core.type.NullaryExpr;
import hu.bme.mit.theta.core.type.booltype.BoolLitExpr;

import static com.google.common.base.Preconditions.checkArgument;
import static hu.bme.mit.theta.core.type.booltype.BoolExprs.Bool;

public final class EnumLitExpr extends NullaryExpr<EnumType> implements LitExpr<EnumType> {

	private final EnumType type;
	private final String value;

	private EnumLitExpr(EnumType type, String value) {
		this.type = type;
		this.value = value;
	}

	public static EnumLitExpr of(EnumType type, String literalName) {
		String value = literalName.contains(".") ? literalName.substring(literalName.indexOf(".") + 1) : literalName;
		checkArgument(type.getValues().contains(value), "Invalid value %s for type %s", value, type.getName());
		return new EnumLitExpr(type, value);
	}

	@Override
	public EnumType getType() {
		return type;
	}

	public String getValue() {
		return value;
	}

	@Override
	public LitExpr<EnumType> eval(Valuation val) {
		return this;
	}

	@Override
	public String toString() {
		return EnumType.makeLongName(type.getName(), value);
	}

	public static BoolLitExpr eq(EnumLitExpr l, EnumLitExpr r) {
		return Bool(l.type.equals(r.type) && l.value.equals(r.value));
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;
		EnumLitExpr that = (EnumLitExpr) o;
		return Objects.equal(type, that.type) && Objects.equal(value, that.value);
	}

	@Override
	public int hashCode() {
		return Objects.hashCode(type, value);
	}
}

File: subprojects/common/core/src/main/kotlin/hu/bme/mit/theta/core/type/enumtype/EnumLitExpr.kt
/*
 *  Copyright 2025 Budapest University of Technology and Economics
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package hu.bme.mit.theta.core.type.enumtype

import hu.bme.mit.theta.core.model.Valuation
import hu.bme.mit.theta.core.type.LitExpr
import hu.bme.mit.theta.core.type.NullaryExpr
import hu.bme.mit.theta.core.type.booltype.BoolExprs.Bool
import hu.bme.mit.theta.core.type.booltype.BoolLitExpr
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
@SerialName("EnumLit")
data class EnumLitExpr(override val type: EnumType, val value: String) :
  NullaryExpr<EnumType>(), LitExpr<EnumType> {

  init {
    require(value in type.values) { "Invalid value $value for type ${type.name}" }
  }

  companion object {

    @JvmStatic
    fun of(type: EnumType, literalName: String): EnumLitExpr {
      val value = EnumType.getShortName(literalName)
      return EnumLitExpr(type, value)
    }

    @JvmStatic
    fun eq(l: EnumLitExpr, r: EnumLitExpr): BoolLitExpr =
      Bool(l.type == r.type && l.value == r.value)

    @JvmStatic
    fun neq(l: EnumLitExpr, r: EnumLitExpr): BoolLitExpr =
      Bool(l.type != r.type || l.value != r.value)
  }

  override fun eval(`val`: Valuation): LitExpr<EnumType> = this

  override fun toString(): String = EnumType.makeLongName(type, value)
}
